src/commands/variables.rs

```diff
@@ -24,53 +43,88 @@ pub async fn command(args: Args, json: bool) -> Result<()> {
-                project_id: linked_project.project.clone(),
-                service_id: service_name.node.id.clone(),
-            },
-            service_name.node.name.clone(),
-        )
-    } else if let Some(ref service) = linked_project.service {
-        let service_name = project
-            .services
-            .edges
+            .find(|service| service.node.name == service_arg || service.node.id == service_arg)
+            .with_context(|| format!("Service '{service_arg}' not found"))?
+            .node
+            .id
+            .to_owned(),
+        // Otherwise if we have a linked service, use that
+        (_, Some(linked_service)) => linked_service,
+        // Otherwise it's a user error
+        _ => bail!(RailwayError::NoServiceLinked),
+    };
+
+    if !args.set.is_empty() {
+        let variables: BTreeMap<String, String> = args
+            .set
             .iter()
-            .find(|edge| edge.node.id == *service)
-            .ok_or_else(|| RailwayError::ServiceNotFound(service.clone()))?;
-        (
-            queries::variables_for_service_deployment::Variables {
-                environment_id: linked_project.environment.clone(),
-                project_id: linked_project.project.clone(),
-                service_id: service.clone(),
-            },
-            service_name.node.name.clone(),
+            .filter_map(|v| {
+                let mut split = v.split('=');
+                let key = split.next()?.trim().to_owned();
+                let value = split.collect::<Vec<&str>>().join("=").trim().to_owned();
+                if value.is_empty() {
+                    None
+                } else {
+                    Some((key, value))
+                }
+            })
+            .collect();
+
+        let fmt_variables = variables
+            .keys()
+            .map(|k| k.bold().to_string())
+            .collect::<Vec<String>>()
+            .join(", ");
+
+        let spinner = indicatif::ProgressBar::new_spinner()
+            .with_style(
+                indicatif::ProgressStyle::default_spinner()
+                    .tick_chars(TICK_STRING)
+                    .template("{spinner:.green} {msg}")
+                    .expect("Failed to set spinner template"),
+            )
+            .with_message(format!("Setting {fmt_variables}..."));
+
+        spinner.enable_steady_tick(Duration::from_millis(100));
+
+        let vars = mutations::variable_collection_upsert::Variables {
+            project_id: linked_project.project.clone(),
+            environment_id,
+            service_id,
+            variables,
+        };
+
+        post_graphql::<mutations::VariableCollectionUpsert, _>(
+            &client,
+            configs.get_backboard(),
+            vars,
         )
-    } else {
-        return Err(RailwayError::NoServiceLinked.into());
-    };
+        .await?;
+
+        spinner.finish_with_message(format!("Set {fmt_variables}"));
+        return Ok(());
```

Comment:
Suggestion: Can this be extracted out to its own function. We could then also have another function to parse and format the variables from the args that is testable

Reply:
I don't think that would be useful. This logic only applies to the --set part.

Reply:
Less so about the usefulness in the command writ large but more so, it's easier from a maintenance perspective to have the function separate so if you or the team adds this elsewhere, we don't have to chase down a bespoke implementation.

Reply:
Fixed